ibis/backends/sql/compiler.py

```diff
@@ -1134,6 +1135,105 @@ def visit_DatabaseTable(
+        # reference that gets dropped in `temporal_join()`. And when it comes to generating
+        # sql (dialects.py) we have `at_time` referencing a stale table with its name
+        # set to an alias (e.g., `t0`) which is not valid anymore.
+        # The following hacky redefinition of `at_time` expression here via `op` is
+        # to "recover" this loss of reference and carry it over to the generators.
+        #
+        # A similar problem with stale table references arises with
+        #
+        # table_right = table_right.at_time(table_right.timestamp_col)
+        #
+        # where `table_right` gets copied in `at_time()` and `at_time` ends up holding on
+        # to the stale table reference. Or with the following example
+        #
+        # expr = table_left.temporal_join(
+        #     table_right.at_time(table_left.timestamp_col),
+        #     predicates=[
+        #         table_left["id"] == table_right["id"],
+        #     ],
+        # )
+        #
+        # where the predicate `table_right["id"]` would be connected to a stale
+        # reference of `table_right` by the time SQL is being generated.
+        #
+        # TODO (mehmet): The above "hack" also does not work because the table name that is
+        # carried over to sqlglot expression `op.at_time.name` holds "table_left" but the
+        # temporal-join sql needs the alias for the most-recent reference for `table_left`.
+        # E.g., For
+        # table_right = table_right.at_time(table_left.timestamp_col)
+        # expr = table_left.temporal_join(
+        #     table_right,
+        #     predicates=[
+        #         table_left["id"] == table_right["id"],
+        #     ],
+        # )
+        #
+        # We get the SQL:
+        # SELECT
+        #   `t1`.`id`,
+        #   `t1`.`bool_col`,
+        #   `t1`.`smallint_col`,
+        #   `t1`.`int_col`,
+        #   `t1`.`timestamp_col`,
+        #   `t3`.`bool_col` AS `bool_col_right`,
+        #   `t3`.`smallint_col` AS `smallint_col_right`,
+        #   `t3`.`int_col` AS `int_col_right`,
+        #   `t3`.`timestamp_col` AS `timestamp_col_right`
+        # FROM `table_left` AS `t1`
+        # JOIN table_right FOR SYSTEM_TIME AS OF `table_left`.timestamp_col AS `t3`
+        #   ON `t1`.`id` = `t3`.`id`
+        #
+        # where `table_left`.timestamp_col is desired to be `t1`.timestamp_col.
+        #
+        # As much as I can see, there are two ways to deal with stale table references:
+        # - Keep track of the table evolution and use the same alias for different copies
+        # of the same table. Might require taking the `alias` dict out of `translate()`
+        # and exposing it for "global use".
+        #   - IMO, this would introduce a lot of tech-debt. May make sense only in the
+        #     presence of other use cases.
+        # - Implement a way to turn off aliasing for temporal joins.
+        #  - This also seems to require additional tech debt.
+
```

Comment:
@kszucs I documented an issue that I could not find a solution for as a long comment in `visit_VersionedDatabaseTable()`. To summarize, I am not able to get `at_time` compile with the right alias due to tables getting copied internally. By the time, SQL is generated for the `nodes`, `at_time` is bound to a "stale" reference of the corresponding table. This leads to generating an incorrect SQL for `at_time`, e.g., `t0.timestamp_col` while it is supposed to be `t2.timestamp_col`. This issue arises in both the initial API and the one you suggested.

I failed resolving/working around this in my initial attempts (though learned quite a bit about the compilation process in Ibis) but will continue looking into this. In the comment, I noted the two approaches that came to my mind but both are quite tedious and likely to be unnecessary. I feel that I am missing something obvious due to my lack of sufficient knowledge of the code base :) Do you have any suggestion for me?

Reply:
Reading the source code more, I might be wrong about my judgement of the table copies "spoiling" the table references. The table reference disconnect for `at_time` seems to be originating from the `JoinTable`'s constructed off of `DatabaseTable` -- see the copied IR graph. Will update here once I get a better understanding.

<img width="1359" alt="image" src="https://github.com/ibis-project/ibis/assets/3505416/36693563-b715-4d78-8998-04d36e6f4d16">

Reply:
We need to wrap the join participant tables in `JoinTable` relations, so that the same table can be added to the join several times. This way we ensure that all tables in the join are distinct and the predicates reference distinct tables. 

This has the drawback that all value expressions referencing the wrapped table, must be replaced with the JoinTable, like `Field(r) => Field(JoinTable(r))`. We call this "dereferencing" and handled by `prepare_predicates()`.
This means that both `table_left` and `table_right` should only be connected to the respective `JoinTable` nodes, but there is an exception in the graph, the `timestamp_col` which indicates that it is not getting dereferenced to the join tables. Most certainly the problem is with the `at_time` here:

```py
link = ops.TemporalJoinLink(how="temporal", table=right, at_time=at_time, predicates=preds)
```

before passing to the join link `at_time` should be dereferenced:

```py
deref_left = dereference_mapping_left(chain)
at_time = at_time.replace(deref_left, filter=ops.Value)
```

Let me know whether that helps or not, then we can iterate with this PR. In the meantime I am trying to have a more detailed look at flink temporal joins.

Reply:
Thanks @kszucs for the detailed explanation. Reading the source code a bit more, I was able to figure that the main reason is indeed the wrapping of participant tables in `JoinTable`. For the initial API, I was able to dereference `at_time` as you explained, but using a much uglier hack 😄:
```python
    def rebase_at_time_on_join_table(self, at_time: Column):
        """This is a "hack" only to demonstrate the challenge with generating
        the correct table alias for `at_time`. Shall be refactored out.

        Creates and returns a new `at_time` such that
        * New `at_time` is created from JoinTable `join_table`
        * `join_table` got created in `relations.py: temporal_join()` from DatabaseTable
        `database_table`.
        * The arg `at_time` got created from `database_table` by the user in the first
        place.

        """
        from ibis.common.graph import Graph

        at_time_source_table = at_time.op().find(finder=ops.DatabaseTable)[0]

        op = self.op()
        graph, dependents = Graph.from_bfs(op).toposort()
        at_time_join_table_op = dependents[at_time_source_table][0]
        at_time_join_table = Table(at_time_join_table_op)
        return at_time_join_table[f"{at_time.op().name}"]
```
I will replace this with your suggestion and share the update here.

For the second API (initiated by your review), I was not able to dereference `at_time` successfully -- my dereferencing "hack" messed up the dereferencing done for the predicates. I will try doing it with `dereference_mapping_left()` and `replace()`, and will share the update here.

Reply:
@kszucs I was able to use `dereference_mapping_left()` and `replace()` to "rebase" the `at_time` field on the left `JoinTable`. I added `TODO` comments for things that I am not certain about.

Reply:
Thanks! Going to have another look hopefully tomorrow.